Update README with overview, setup guide, quickstart and add practical examples

rtweera opened the pull request:
## Purpose
This PR updates the module documentation by adding an overview, setup guide, quick start, and examples to streamline user onboarding.

## Examples

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

rtweera committed:
Add overview in README.md

Add overview section in documentation with links to HubSpot API
documentation for video conferrencing settings module, HubSpot developer
portal,and HubSpot REST API documentation.

This gives a brief overview of the module and the resources it provides.

rtweera committed:
Add setup guide in README.md

Add setup guide section in documentation with screenshots for setting up
the video conferrencing settings module.

This gives a step-by-step easy to follow guide on how to get
authentication tokens from the HubSpot, how to set up the test apps and
developer account in HubSpot in order to use this module.

rtweera committed:
Merge branch 'main' into 'documentation'

Bringing the documentation branch up to date with the latest changes from the main branch.# Please enter a commit message to explain why this merge is necessary,

rtweera committed:
Add quick start guide in README.md

Add steps and code segments to setup the connector and use it to create
or update meeting settings.

rtweera committed:
Merge branch 'main' to 'documentation' for syncing

rtweera committed:
Add examples

Add examples for changing settings when operating and shutting down of
external video conferencing applications.

rtweera committed:
Add examples to README.md

Link the README.md file's example section to the examples directory and
list the examples in the README.md file.

rtweera committed:
Fix image exposing sensitive information

Hide part of an image exposing an App ID.

rtweera committed:
Change the generated examples link in README.md

Generated link for examples directory uses an absolute path bound to
original repository. This commit changes the link to a relative path
to the examples directory.

rtweera committed:
Update ./examples/README.md

- Add examples list and links to the README.md file.
- Add prerequisites for using the examples.

rtweera committed:
Fix incorrect link in examples' READMEs

- Update setup guide link in operate-conference-service example
- Update setup guide link in close-conference-service example

rtweera committed:
Update alt texts for images in ./README.md

Update alt texts to match the actual image.

rtweera committed:
Fill ballerina/Module.md and ballerina/Package.md

Fill ballerina/Module.md and ballerina/Package.md files with overview,
setup guide, quickstart, and examples.

rtweera committed:
Merge branch 'main' to 'documentation'

rtweera committed:
Modify README.md to correct formatting

- Correct formatting in README.md
- Update image in step 3.1
- Change image links from relative addresses to absolute addresses

rtweera committed:
Change example links in README.md

Change to absolute links instead of the relative links in the README.md
file.

rtweera committed:
Modify links in examples/README.md

Update the links in the examples/README.md file to point to the absolute
repository path of the examples.

rtweera committed:
Modify README.md in specific examples

Change links to absolute paths in examples/operate-conference-service/README.md
and examples/close-conference-service/README.md.

Change import to a simpler name.

Modify example descriptions.

rtweera committed:
Change Module.md and Package.md to README.md

Remove both Module.md, Package.md and add README.md. Change is done due
to the new structure of Ballerina Swan Lake Update 11.

Nuvindu commented on docs/resources/step-2-1.png:
Shall we rename these files with their respective case. eg: (build-public-app.png) ?
And make sure to apply those changes in the docs as well

Nuvindu commented on examples/README.md:
```diff
@@ -2,13 +2,38 @@
-[//]: # (TODO: Add examples)
-1. 
-2. 
+1. [Operate Conference Service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/operate-conference-service/)
+2. [Close Conference Service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/close-conference-service/)
```
```suggestion
1. [Operate conference service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/operate-conference-service)
2. [Close conference service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/close-conference-service)
```

Nuvindu commented on examples/close-conference-service/.devcontainer.json:
Let's remove this file

rtweera replied:
done

Nuvindu commented on examples/close-conference-service/.gitignore:
Remove this file too.

Nuvindu commented on examples/close-conference-service/Ballerina.toml:
```diff
@@ -0,0 +1,14 @@
+
+[build-options]
+observabilityIncluded = true
+
+[[dependency]]
+org = "ballerinax"
+name = "hubspot.crm.extensions.videoconferencing"
+version = "1.0.0"
+repository="local"
```
```suggestion

```

Nuvindu commented on examples/close-conference-service/README.md:
```diff
@@ -0,0 +1,67 @@
+# [Close Conference Service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/close-conference-service/)
```
```suggestion
# Remove saved settings in a video conferencing service
```

rtweera replied:
done

Nuvindu commented on examples/close-conference-service/README.md:
````diff
@@ -0,0 +1,67 @@
+    appId = <App ID as an int>
+    ```
+
+4. **`hubspot.crm.extensions.videoconferencing` module:** Import the `ballerinax/hubspot.crm.extensions.videoconferencing` module into your Ballerina project and configure it with the obtained credentials.
+
+    ```ballerina
+    import ballerinax/hubspot.crm.extensions.videoconferencing as hubspot;
+
+    configurable string hapikey = ?;
+    configurable int appId = ?;
+
+    final int:Signed32 appIdSigned32 = <int:Signed32>appId;
+
+    final hubspot:ApiKeysConfig apiKeysConfig = {
+        hapikey: hapikey
+    };
+
+    final hubspot:Client baseClient = check new (apiKeysConfig);
+    ```
````
Actually, this is not a prerequisite. We can remove this

rtweera replied:
Okay sure

Nuvindu commented on examples/operate-conference-service/.devcontainer.json:
We can remove this file

rtweera replied:
done

Nuvindu commented on examples/operate-conference-service/.gitignore:
Remove this file as well

rtweera replied:
done

Nuvindu commented on examples/operate-conference-service/Ballerina.toml:
```diff
@@ -0,0 +1,14 @@
+
+[build-options]
+observabilityIncluded = true
+
+[[dependency]]
+org = "ballerinax"
+name = "hubspot.crm.extensions.videoconferencing"
+version = "1.0.0"
+repository="local"
```
```suggestion

```

Nuvindu commented on examples/operate-conference-service/README.md:
```diff
@@ -0,0 +1,67 @@
+# [Operate Conference Service](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.extensions.videoconferencing/tree/main/examples/operate-conference-service/)
```
```suggestion
# Create and manage settings in a video conference service
```

rtweera replied:
done

Nuvindu commented on examples/operate-conference-service/main.bal:
```diff
@@ -0,0 +1,105 @@
+
+import ballerina/io;
+import ballerina/lang.runtime;
+import ballerinax/hubspot.crm.extensions.videoconferencing as hsvideoconferencing;
```
Format this file

Nuvindu commented on examples/operate-conference-service/main.bal:
```diff
@@ -0,0 +1,105 @@
+final hsvideoconferencing:ApiKeysConfig apiKeysConfig = {
+    hapikey: hapikey
+};
+final hsvideoconferencing:Client hsVideoConferencing = check new (apiKeysConfig);
```
We can use `hubspot` as the client variable name

rtweera replied:
done

rtweera committed:
Change image names in docs/resources

Change image names to be more descriptive of the content they contain.
Reflected the changes in ./README.md and ./ballerina/README.md files.

rtweera committed:
Format ./examples/README.md

Update link texts in title case to sentence case.

rtweera committed:
Format ./ballerina/tests/test.bal

rtweera committed:
Remove gitignore, devcontainer files in examples

rtweera committed:
Remove unnecessary prerequisites from README.md

Remove the hubspot module import part in the prerequisites for READMEs
in the examples. This is because they are not prerequisites for following
the examples.

rtweera committed:
Rename client object to a simpler one

Instead of the lengthy client name `hsVideoConferencing`, change it to
`hubspot` for easier readability.

rtweera committed:
Remove local dependency from Ballerina.toml files

Remove in the ./examples/**

Module will be published to ballerina central hence local dependencies
are not needed to run the examples.

rtweera committed:
Change example names to be more descriptive

Update ./examples/open-conference-service/README.md and
./examples/close-conference-service/README.md to have more descriptive
names (i.e. titles) for the examples.

Reflect these changes in the README.md, ballerina/README.md,
./examples/README.md files.

rtweera committed:
Fix spelling errors in ./examples/**/README.md

ThisaruGuruge commented on examples/README.md:
```diff
@@ -2,13 +2,21 @@
 ## Prerequisites
 
-[//]: # (TODO: Add prerequisites)
+1. **Ballerina:** Download and install Ballerina from [here](https://ballerina.io/downloads/).
```
```suggestion
1. Ballerina Swan Lake Update 11 (2201.11.0)
```

rtweera replied:
Done

ThisaruGuruge commented on examples/close-conference-service/Ballerina.toml:
```diff
@@ -0,0 +1,8 @@
+distribution = "2201.11.0"
+
+[build-options]
+observabilityIncluded = true
```
```suggestion
observabilityIncluded = true

```

rtweera replied:
Done

ThisaruGuruge commented on examples/close-conference-service/README.md:
```diff
@@ -0,0 +1,50 @@
+# Remove saved settings for a video conferencing service
+
+This example demonstrates how to use the `HubSpot CRM Video conference connector` to remove saved settings related to an external video conferencing service. This covers the removal of settings for notifying the external video conferencing service about meetings scheduled in the HubSpot web site by users. This allows to shut down the external video conferencing application safely without getting  ...
```
```suggestion
This example demonstrates how to use the `HubSpot CRM Video conference connector` to remove saved settings related to an external video conferencing service. This covers the removal of settings for notifying the external video conferencing service about meetings scheduled in the HubSpot CRM by users. This allows to shut down the external video conferencing application safely without getting into any errors on HubSpot end or HubSpot making unnecessary requests to non-existing URLs (which were previously held by the external video conferencing application).
```

rtweera replied:
Done

ThisaruGuruge commented on examples/close-conference-service/main.bal:
```diff
@@ -0,0 +1,81 @@
+
+    // Step 2: Get the current settings of the external video conferencing app from HubSpot App.
+    runtime:sleep(60);
+    hsvideoconferencing:ExternalSettings|error currentSettings = hubspot->/[appIdSigned32]();
+    if currentSettings !is hsvideoconferencing:ExternalSettings {
+        panic error("Error getting settings");
+    }
```
```suggestion
    hsvideoconferencing:ExternalSettings currentSettings = check hubspot->/[appIdSigned32]();
```

rtweera replied:
Done

ThisaruGuruge commented on examples/close-conference-service/main.bal:
```diff
@@ -0,0 +1,81 @@
+
+    // Step 3: Remove the updateMeetingUrl from the settings.
+    hsvideoconferencing:ExternalSettings updatedSettings = {
+        createMeetingUrl: currentSettings.createMeetingUrl,
+        deleteMeetingUrl: currentSettings?.deleteMeetingUrl
```
```suggestion
        ...currentSettings
```

rtweera replied:
Here `currentSettings` (which is the result received from a `get`) is an object like this:

```ballerina
{
   createMeetingUrl: "some url",
   updateMeetingUrl: "some url2",
   deleteMeetinUrl: "some url3"
}
```

In such case should we do `...currentSettings` or keep it as it was before?

rtweera replied:
`currentSettings` contains the `updateMeetingUrl` field but we want to remove it for the next object that we are setting.

rtweera replied:
Decided to keep it as is.

ThisaruGuruge commented on examples/close-conference-service/main.bal:
```diff
@@ -0,0 +1,81 @@
+    };
+
+    // Step 4: Verify whether the updateMeetingUrl is removed from HubSpot App.
+    hsvideoconferencing:ExternalSettings|error updatedSettingsResponse = hubspot->/[appIdSigned32].put(updatedSettings);
```
```suggestion
    hsvideoconferencing:ExternalSettings updatedSettingsResponse = check hubspot->/[appIdSigned32].put(updatedSettings);
```

rtweera replied:
Done

ThisaruGuruge commented on examples/close-conference-service/main.bal:
```diff
@@ -0,0 +1,81 @@
+    // Step 2: Verify whether all the settings are deleted from HubSpot App.
+    if deleteResponse.statusCode == 204 {
+        io:println("All settings deleted successfully for scenario 2");
+    } else {
+        io:println("Error deleting all settings for scenario 2");
+    }
```
```suggestion
    }
    io:println("Error deleting all settings for scenario 2");
```

rtweera replied:
This is the code segment mentioned in the above suggestion. I am checking for the `statusCode` of the API response. In that case, we have to keep the `else` part as it is right? 

```ballerina
if deleteResponse.statusCode == 204 {
    io:println("All settings deleted successfully for scenario 2");
} else {
    io:println("Error deleting all settings for scenario 2");
}
```

rtweera replied:
Decided to keep as is.

ThisaruGuruge commented on examples/operate-conference-service/main.bal:
```diff
@@ -0,0 +1,105 @@
+configurable string hapikey = ?;
+configurable int appId = ?;
+
+final int:Signed32 appIdSigned32 = <int:Signed32>appId;
```
```suggestion
final int:Signed32 appIdSigned32 = check appId.ensureType();
```

rtweera replied:
Done

ThisaruGuruge commented on examples/operate-conference-service/main.bal:
```diff
@@ -0,0 +1,105 @@
+final int:Signed32 appIdSigned32 = <int:Signed32>appId;
+
+final hsvideoconferencing:ApiKeysConfig apiKeysConfig = {
+    hapikey: hapikey
```
```suggestion
    hapikey
```

rtweera replied:
Done

rtweera committed:
Modify changes to documentation

Change formatting to be more consistent and easier to read.
Change some titles to be more concise.
Change example code segments to reflect the best practices in Ballerina.

rtweera committed:
Update example bal files with improved code

- Replace cast with ensureType() for better error handling.
- Replace hardcoded value checks with object field checks.
- Replace error type with check for better readability.
- Add configurable delay.

rtweera committed:
Improve test.bal with better code snippets

- Add delay back
- Use ensureType() for better error handling
- Add shorthand for apiKeysConfig fields

rtweera committed:
Update README.md files in ./examples

- Change Ballerina prerequsite in close-conference-service example
- Add description to example links in ./examples/README.md

rtweera committed:
Remove icon field in ballerina.toml

`icon` field is not needed.

rtweera committed:
Add keywords to the Ballerina.toml file

rtweera committed:
Format examples/README.md

Remove additional newline in the middle.

rtweera committed:
Change image URLs to add refs/heads in READMEs

Without refs/heads, the image URLs could break sometimes.
